sway-lsp/src/core/token.rs

```diff
@@ -91,3 +62,113 @@ pub enum SymbolKind {
     TypeParameter,
     Unknown,
 }
+
+#[derive(Debug, Clone)]
+pub enum TypeDefinition {
+    TypeId(TypeId),
+    Ident(Ident),
+}
+
+#[derive(Debug, Clone)]
+pub struct Token {
+    pub parsed: AstToken,
+    pub typed: Option<TypedAstToken>,
+    pub type_def: Option<TypeDefinition>,
+    pub kind: SymbolKind,
+}
+
+impl Token {
+    /// Create a new token with the given SymbolKind.
+    /// This function is intended to be used during traversal of the
+    /// `ParseProgram` AST.
+    pub fn from_parsed(token: AstToken, kind: SymbolKind) -> Self {
+        Self {
+            parsed: token,
+            typed: None,
+            type_def: None,
+            kind,
+        }
+    }
+
+    /// Return the `Ident` of the declaration of the provided token.
+    pub fn declared_token_ident(&self, type_engine: &TypeEngine) -> Option<Ident> {
```

Comment:
Is the need to pass `type_engine` because the compiler no longer uses lazy statics (this tells it which session we're using)?

Reply:
Yeah pretty much. Internally we call `ident_of_type_id` which converts from a `TypeId` to a `TypeInfo`. There is a method on the type_engine that does this `type_engine.look_up_type_id(*type_id)`. So in order to have the `TypeId` return the correct `TypeInfo`, we need to pass in the type_engine associated with the current session.